---
title: "05.SNPs_analysis"
author: "Bosco Gracia Alvira"
date: "2024-07-25"
output: html_document
---

```{r setup}
knitr::opts_knit$set(root.dir = "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Microbiome_pangenomic_analysis/data/Lactiplantibacillus_plantarum/SNPs_analysis")
knitr::opts_chunk$set(echo = FALSE)
library(lattice)
library(data.table)
library(Biostrings)
library(ggplot2)
library(dplyr)
library(tidyverse)
library(stringr)
library(gtools)

```



```{r Loading the data}

# Paths to the files that we want to import
data_path <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Microbiome_pangenomic_analysis/data/Lactiplantibacillus_plantarum/SNPs_analysis"
metadata_path <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Microbiome_pangenomic_analysis/data/Lactiplantibacillus_plantarum/Anvio/Anvio_misc.tsv"
visuals_path <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Microbiome_pangenomic_analysis/visuals"

metadata <- fread(metadata_path, header=T)
pca <- read.table(paste0(data_path,"/Lactiplantibacillus_plantarum.eigenvec"),sep=" ",header=F)

colnames(pca)[1] <- "name"
pca[,2] <- NULL

pi.all <- read.table(paste0(data_path,"/Lactiplantibacillus_plantarum_10kb.windowed.pi"),header=T)



```


```{r PCA of the SNPs}

pca_merged <- merge(pca, metadata, by = "name")

ggplot(pca_merged, aes(x = V3, y = V4, color = Generation)) +
  geom_point(aes(shape=Temperature)) +
  labs(x = "PC1", y = "PC2", color = "Generation", shape="Temperature") +
  theme_minimal()
ggsave(filename = paste0(visuals_path,"/Lpla_SNPs_PCA.png"), width = 6, height = 4, dpi = 300)

```



```{r Nucleotide diversity}






```




```{r Tajima's D}








```

