---
title: "05.SNPs_analysis"
author: "Bosco Gracia Alvira"
date: "2024-07-25"
output: html_document
---

```{r setup}
knitr::opts_knit$set(root.dir = "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Microbiome_pangenomic_analysis/data/Lactiplantibacillus_plantarum/SNPs_analysis")
knitr::opts_chunk$set(echo = FALSE)
library(lattice)
library(data.table)
library(Biostrings)
library(ggplot2)
library(dplyr)
library(tidyverse)
library(stringr)
library(gtools)

```



```{r Loading the data}

# Paths to the files that we want to import
data_path <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Microbiome_pangenomic_analysis/data/Lactiplantibacillus_plantarum/SNPs_analysis"
metadata_isolates_path <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Microbiome_pangenomic_analysis/data/Lactiplantibacillus_plantarum/Anvio/Anvio_misc.tsv"
visuals_path <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Microbiome_pangenomic_analysis/visuals"

# freq is the table that comes from the vcf query that contains allele frequencies for each sample
freq <- fread(paste0(data_path,"/Lactiplantibacillus_plantarum.freq"),sep="\t",header=T)

# Make the headers pretty :)
colnames(freq) <- gsub("^#", "", colnames(freq))          # Remove "#"
colnames(freq) <- gsub("^\\[.*?\\]", "", colnames(freq))  # Remove anything between brackets
colnames(freq) <- gsub(":AD$", "", colnames(freq))        # Remove ":AD"

# Melt the dataframe to long format
freq_long <- data.table::melt(freq, id.vars = colnames(freq)[1:4], variable.name = "Sample", value.name = "AF")

# Split the AF column into Reference and alternative using "," as splitting element
freq_long <- freq_long[, c("RO", "AO") := tstrsplit(AF, ",")]

# Turn the new columns numeric
freq_long <- freq_long[, `:=`(RO = as.numeric(RO), AO = as.numeric(AO))]

# Calculate the summed allele depth (SAD) and relative frequencies from absolute number of counts
freq_long <- freq_long[, "SAD" := (RO + AO)]
freq_long <- freq_long[, `:=`(RO = RO / (RO + AO), AO = AO / (RO + AO))]

# Create a new column with unique position information
freq_long <- freq_long %>%
  mutate(CHROM_POS = paste(CHROM, POS, sep = "_"))

# Make the table wide again
freq_wide <- freq_long %>%
  select(Sample, CHROM_POS, RO) %>%  # Select relevant columns
  pivot_wider(names_from = CHROM_POS, values_from = RO) %>%  # Pivot to wide format
  column_to_rownames(var = "Sample")  # Set Sample as row names

# Remove problematic samples hF130
freq_wide <- freq_wide[rownames(freq_wide) != "hF130", ]

# Remove the columns (SNPs) that are missing in at least 1 sample
freq_wide <- freq_wide %>%
  select_if(~ !any(is.nan(.)))

# Transform data to arcosin
freq_asin <- 2*asin(sqrt(freq_wide))

pca.res <- prcomp(freq_asin, retx=TRUE, center = FALSE, scale. =FALSE)

#Visualize eigenvalues (scree plot).
  #print(fviz_eig(pca.res))
  
  #Show the percentage of variances explained by each principal component.
  #print(fviz_contrib(pca.res, choice = "var", axes = 1:2, top = 10))
  
  #if mean.point = TRUE group means are added for each group, specified by coloring_factor
  print(fviz_pca_ind(pca.res,
                     repel = TRUE,   # Avoid text overlapping
                     mean.point = FALSE,
                     addEllipses=FALSE)) # Default parameters for ellipses



result <- freq_long %>%
  group_by(Sample) %>%
  summarise(
    SFS = list(table(AF)),
    Median_RO = median(RO)
  )


pca <- read.table(paste0(data_path,"/Lactiplantibacillus_plantarum.eigenvec"),sep=" ",header=F)
colnames(pca)[1] <- "name"
pca[,2] <- NULL

# How much variance is explained by each PC?
eigenval <- scan(paste0(data_path,"/Lactiplantibacillus_plantarum.eigenval"))
variance_explained <- eigenval / sum(eigenval)
percentage_variance_explained <- variance_explained * 100


pca_5 <- read.table(paste0(data_path,"/Lactiplantibacillus_plantarum_5x.eigenvec"),sep=" ",header=F)
colnames(pca_5)[1] <- "name"
pca_5[,2] <- NULL

pca_10 <- read.table(paste0(data_path,"/Lactiplantibacillus_plantarum_10x.eigenvec"),sep=" ",header=F)
colnames(pca_5)[1] <- "name"
pca_10[,2] <- NULL

metadata_isolates <- fread(metadata_isolates_path, header=T)
metadata_isolates$Source <- "Isolate"

           
metadata_pools <- as.data.frame(pca$name[grepl("^cF|^hF", pca$name)])
colnames(metadata_pools) <- "name"
metadata_pools <- metadata_pools %>%
  mutate(
    Temperature = case_when(
      str_detect(name, "F0") ~ "BASE",
      str_detect(name, "^h") ~ "HOT",
      str_detect(name, "^c") ~ "COLD"),
    Generation = str_extract(name, "F\\d+"),
    Source = "Pool")

metadata <- rbind(metadata_isolates,metadata_pools)

# Create a colour palette for our temperatures
temp_palette <- c("HOT" = "red", "BASE" = "grey", "COLD" = "lightblue")

```


```{r PCA of the SNPs}

pca_merged <- merge(pca, metadata, by = "name")
pca_pool <- subset(pca_merged, pca_merged$Source=="Pool")
pca_isolate <- subset(pca_merged, pca_merged$Source=="Isolate")


ggplot() +
  geom_point(data = pca_isolate, aes(x = V3, y = V4, color = Temperature, shape = Source), size = 1, alpha = 0.4) +
  geom_point(data = pca_pool, aes(x = V3, y = V4, color = Temperature, shape = Source), size = 2) +
  geom_text(data = pca_pool, aes(x = V3, y = V4, label = name, color = Temperature), size = 2, hjust = -0.5) +
  labs(
  x = paste0("PC1 (", round(percentage_variance_explained[1], 2), "% variance)"),
  y = paste0("PC2 (", round(percentage_variance_explained[2], 2), "% variance)"),
  color = "Temperature", shape = "Source") +
  scale_colour_manual(values = c("HOT" = "red", "BASE" = "grey", "COLD" = "lightblue", "COLD-CONSTANT" = "blue")) +
  theme_minimal()
ggsave(filename = paste0(visuals_path,"/Lpla_SNPs_PCA.png"), width = 6, height = 4, dpi = 300)



ggplot(pca_merged, aes(x = V3, y = V4, color = Temperature)) +
  #geom_point(aes(shape=Source)) +
  geom_text(aes(label = name), size = 1) +
  labs(x = "PC1", y = "PC2", color = "Temperature", shape="Source", title = "PCA of L. plantarum") +
  scale_colour_manual(values = c("HOT" = "red", "BASE" = "grey", "COLD" = "lightblue", "COLD-CONSTANT" = "blue")) +
  theme_minimal()
ggsave(filename = paste0(visuals_path,"/Lpla_SNPs_PCA.png"), width = 6, height = 4, dpi = 300)

# This gives a sparser PCA, but the dots are falsely placed!!!
jitter_position <- position_jitter(width = 0.05, height = 0.05)

ggplot(pca_merged, aes(x = V3, y = V4, color = Temperature)) +
  geom_point(aes(shape=Source), size = 1, alpha = 0.6, , position = jitter_position) +
  geom_text(aes(label = name), size = 1, alpha = 0.6, , position = jitter_position, vjust = -1, hjust = 0.5) +
  labs(x = "PC1", y = "PC2", color = "Temperature", shape="Source", title = "PCA of L. plantarum (jittered)") +
  scale_colour_manual(values = c("HOT" = "red", "BASE" = "grey", "COLD" = "lightblue", "COLD-CONSTANT" = "blue")) +
  theme_minimal()

pca_5_merged <- merge(pca_5, metadata, by = "name")

ggplot(pca_5_merged, aes(x = V3, y = V4, color = Temperature)) +
  geom_point(aes(shape=Source)) +
  labs(x = "PC1", y = "PC2", color = "Temperature", shape="Source", title = "PCA of L. plantarum 5x") +
  scale_colour_manual(values = c("HOT" = "red", "BASE" = "grey", "COLD" = "lightblue", "COLD-CONSTANT" = "blue")) +
  theme_minimal()

pca_10_merged <- merge(pca_5, metadata, by = "name")

ggplot(pca_10_merged, aes(x = V3, y = V4, color = Temperature)) +
  geom_point(aes(shape=Source)) +
  labs(x = "PC1", y = "PC2", color = "Temperature", shape="Source", title = "PCA of L. plantarum 10x") +
  scale_colour_manual(values = c("HOT" = "red", "BASE" = "grey", "COLD" = "lightblue", "COLD-CONSTANT" = "blue")) +
  theme_minimal()

```



```{r Nucleotide diversity}






```




```{r Tajima's D}








```

