---
title: "Lac_phylogeny"
author: "Bosco Gracia Alvira"
date: "2024-10-16"
output: html_document
---

```{r Load libraries}

#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
BiocManager::install("ggtree")
BiocManager::install("treeio")
BiocManager::install("Biostrings")
install.packages("dplyr")
install.packages("readr")

library(tidyverse) ; packageVersion("tidyverse")
library(ggtree) ; packageVersion("ggtree")
library(treeio) ; packageVersion("treeio")
library(readr) ; packageVersion("readr")
library(phyloseq) ; packageVersion("phyloseq")
library(vegan) ; packageVersion("vegan")
library(dplyr) ; packageVersion("dplyr")
library(RColorBrewer) ; packageVersion("RColorBrewer")
library(dendextend) ; packageVersion("dendextend")
library(ape) ; packageVersion("ape")
library(data.table) ; packageVersion("data.table")

```

```{r Set paths}

tree_path <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Microbiome_pangenomic_analysis/data/Lactiplantibacillus_plantarum/Phylogeny/parsnp/parsnp.tree"
metadata_path <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Microbiome_pangenomic_analysis/data/Lactiplantibacillus_plantarum/Phylogeny/metadata.tsv"
visuals_path <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Microbiome_pangenomic_analysis/visuals"

```

```{r}

# Load the metadata file
metadata <- fread(metadata_path)

# Load the representative genomes tree based on GTDB-Tk MSA
tree <- read.tree(tree_path)
tree <- ladderize(tree, right = FALSE)

# Remove the .fa suffix from the tip labels
tree$tip.label <- sub("\\.fa$", "", tree$tip.label)

# Modify the outgroup tip label
tree$tip.label[tree$tip.label == "Lpla_ref.gbff.fna"] <- "Reference"

# Re-root the tree to the outgroup
new_root_node <- which(tree$tip.label == "NIZO1840")

# Re-root the tree
tree <- root(tree, new_root_node)

# Convert the tree to a tibble
tree_merged <- as_tibble(tree)

# Merge the tree data with metadata
tree_merged <- tree_merged %>%
    left_join(metadata, by = c("label" = "Strain"))

# Build the tree and save it in a figure
# Open a graphics device (e.g., PNG)
png(filename = "Lac_phylogeny.png")


tree_plot <- ggtree(tree, size = 0.3) %<+% tree_merged + 
                geom_tippoint(aes(color = Source), size = 0.5) +
                geom_tiplab(aes(color = Source), size = 2) +
                scale_color_manual(values = setNames(metadata$Colour, metadata$Source)) +
                theme_tree2()

tree_plot

ggsave(filename = "Lac_phylogeny.png", plot = tree_plot, path = visuals_path, width = 20, height = 6, dpi = 300)

# Close the graphics device
dev.off()

```

