---
title: "Lac_strains_mutations"
author: "Bosco Gracia Alvira"
date: "2025-02-24"
output: html_document
---

```{r Load libraries}

# Standard table management
library(data.table) ; packageVersion("data.table")
library(tidyverse) ; packageVersion("tidyverse")
library(readr) ; packageVersion("readr")
library(dplyr) ; packageVersion("dplyr")
library(reshape2) ; packageVersion("reshape2")

# Load genomic data
library(Biostrings) ; packageVersion("Biostrings")

# Ordination of the pairwise distances
library(factoextra) ; packageVersion("factoextra")

# Plot graphs
library(ggplot2) ; packageVersion("ggplot2")
library(igraph) ; packageVersion("igraph")
library(ggraph) ; packageVersion("ggraph")
library(tidygraph) ; packageVersion("tidygraph")
library(pheatmap) ; packageVersion("pheatmap")

```


```{r Set paths}

snps_path <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Microbiome_pangenomic_analysis/data/Lactiplantibacillus_plantarum/Pangenomic_strains/snp-dists/pairwise_snps.tsv"

snps_long_path <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Microbiome_pangenomic_analysis/data/Lactiplantibacillus_plantarum/Pangenomic_strains/snp-dists/pairwise_snps_long.tsv"

aln_path <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Microbiome_pangenomic_analysis/data/Lactiplantibacillus_plantarum/Pangenomic_strains/snp-sites/Lpla.snp_sites.aln"

tree_path <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Microbiome_pangenomic_analysis/data/Lactiplantibacillus_plantarum/Pangenomic_strains/snp-sites/Lpla.snp_sites.aln"

accessory_path <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Microbiome_pangenomic_analysis/data/Lactiplantibacillus_plantarum/Pangenomic_strains/roary/gene_presence_absence.Rtab"
  
metadata_path <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Microbiome_pangenomic_analysis/data/Lactiplantibacillus_plantarum/Pangenomic_strains/metadata.tsv"

visuals_path <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Microbiome_pangenomic_analysis/visuals"

```

```{r Load the data and subset}

# Load the metadata with info from each isolate
metadata <- fread(metadata_path) %>% 
                        mutate(GenPop = paste0(Generation,"_",Population),
                               GenPop = gsub(" ", "", GenPop))

row.names(metadata) <- metadata$Strain
setkey(metadata, Strain)

# Load the snps files
snps <- fread(snps_path) %>% column_to_rownames(var = colnames(.)[1])
snps_long <- fread(snps_long_path) %>% setNames(c("Genomes1","Genomes2","SNPs"))

# Subset the data to within-strain isolates
isolates_S239 <- metadata %>%
  filter(Genotype == "S239") %>%
  pull(Strain)

snps_S239 <- snps %>%
  filter(row.names(snps) %in% isolates_S239) %>%
  select(all_of(isolates_S239))

snps_S239_long <- snps_S239 %>% 
                      rownames_to_column(var = "Genome1") %>% 
                      pivot_longer(cols = -Genome1, names_to = "Genome2", values_to = "SNPs") %>% 
                      left_join(metadata, by = c("Genome1" = "Strain")) %>%
                      dplyr::rename(Generation1 = Generation) %>% 
                      left_join(metadata, by = c("Genome2" = "Strain")) %>%
                      dplyr::rename(Generation2 = Generation) %>% 
                      mutate(Same_Generation = ifelse(Generation1 == Generation2, "Yes", "No"))

density_S239 <- ggplot(snps_S239_long, aes(x = SNPs)) +
                  geom_density(fill = "red", alpha = 0.5) +
                  labs(title = "Density Plot (hot)", x = "Value", y = "Density") +
                  theme_minimal()

box_S239 <- ggplot(snps_S239_long, aes(x = Same_Generation, y = SNPs)) +
                  geom_boxplot(colour ="red") +
                  labs(title = "Pairwise SNPs between isolates (hot)", x = "Same generation comparisons", y = "Pairwise SNPs") +
                  theme_minimal()
density_S239
box_S239

isolates_S103 <- metadata %>%
  filter(Genotype == "S103") %>%
  pull(Strain)

snps_S103 <- snps %>%
  filter(row.names(snps) %in% isolates_S103) %>%
  select(all_of(isolates_S103))

snps_S103_long <- snps_S103 %>% 
                      rownames_to_column(var = "Genome1") %>% 
                      pivot_longer(cols = -Genome1, names_to = "Genome2", values_to = "SNPs") %>% 
                      left_join(metadata, by = c("Genome1" = "Strain")) %>%
                      dplyr::rename(Generation1 = Generation) %>% 
                      left_join(metadata, by = c("Genome2" = "Strain")) %>%
                      dplyr::rename(Generation2 = Generation) %>% 
                      mutate(Same_Generation = ifelse(Generation1 == Generation2, "Yes", "No"))

density_S103 <- ggplot(snps_S103_long, aes(x = SNPs)) +
                  geom_density(fill = "blue", alpha = 0.5) +
                  labs(title = "Density Plot (cold)", x = "Value", y = "Density") +
                  theme_minimal()

box_S103 <- ggplot(snps_S103_long, aes(x = Same_Generation, y = SNPs)) +
                  geom_boxplot(colour ="blue") +
                  labs(title = "Pairwise SNPs between isolates (cold)", x = "Same generation comparisons", y = "Pairwise SNPs") +
                  theme_minimal()
density_S103
box_S103

```

```{r Ordination of the isolates by their SNPs distances}

# Define the colors for the generations
list_colours <- list(
  GenPop = c("F1_Florida" = "grey",
             "F57_SouthAfrica" = "green",
             "F89_Florida" = "lightblue",
             "F109_Florida" = "blue",
             "F175_Florida" = "red",
             "F202_Portugal" = "orange",
             "F262_Florida" = "darkred")
)

df_colours <- list_colours$GenPop

# Convert the symmetric table to a distance matrix
dist_S239 <- as.dist(snps_S239)

# Perform PCA on the distance matrix
pca_S239 <- prcomp(as.matrix(dist_S239), scale. = TRUE)
pca_S239_merged <- as.data.frame(pca_S239$x) %>% 
                                rownames_to_column(var = "Strain") %>% 
                                left_join(metadata, by = join_by(Strain))

pca_S239_plot <- ggplot(pca_S239_merged, aes(x=PC1,y=PC2, colour=GenPop)) + 
                    geom_point(aes(shape=Population)) +
                    labs(title = "Ordination of S239 isolates") +
                    scale_colour_manual(values = df_colours) +
                    theme_minimal()
pca_S239_plot

ggsave(filename = "Lac_mutations_PCA_S239.png", plot = pca_S239_plot, path = visuals_path, width = 6, height = 4, dpi = 300)

# Convert the symmetric table to a distance matrix
dist_S103 <- as.dist(snps_S103)

# Perform PCA on the distance matrix
pca_S103 <- prcomp(as.matrix(dist_S103), scale. = TRUE)
pca_S103_merged <- as.data.frame(pca_S103$x) %>% 
                                rownames_to_column(var = "Strain") %>% 
                                left_join(metadata, by = join_by(Strain))

pca_S103_plot <- ggplot(pca_S103_merged, aes(x=PC1,y=PC2, colour=GenPop)) + 
                    geom_point(aes(shape=Population)) +
                    labs(title = "Ordination of S103 isolates") +
                    scale_colour_manual(values = df_colours) +
                    theme_minimal()
pca_S103_plot

ggsave(filename = "Lac_mutations_PCA_S103.png", plot = pca_S103_plot, path = visuals_path, width = 6, height = 4, dpi = 300)

```


```{r Heat map S239}
# S239

# Create the annotation data frame
annotation_S239 <- metadata %>% 
                    filter (Genotype == "S239") %>% 
                    select(c(Strain,GenPop))

rownames(annotation_S239) <- annotation_S239$Strain
annotation_S239$Strain <- NULL

matrix_S239 <- as.matrix(snps_S239)
nums_S239 <- apply(matrix_S239, c(1, 2), function(x) sprintf("%d", x))

# Plot the heatmap using pheatmap
pheatmap(matrix_S239, 
         cluster_rows = TRUE, 
         cluster_cols = TRUE, 
         display_numbers = nums_S239, 
         color = colorRampPalette(c("yellow", "purple"))(100),
         main = "SNPs heatmap of the strain S239",
         annotation_col = annotation_S239,
         annotation_row = annotation_S239,
         annotation_colors = list_colours)

png(paste0(visuals_path,"/Lac_mutations_heat_S239.png"), width = 4000, height = 3000, res = 300)
pheatmap(matrix_S239, 
         cluster_rows = TRUE, 
         cluster_cols = TRUE, 
         display_numbers = nums_S239, 
         color = colorRampPalette(c("yellow", "purple"))(100),
         main = "SNPs heatmap of the strain S239",
         annotation_col = annotation_S239,
         annotation_row = annotation_S239,
         annotation_colors = list_colours)
dev.off()

```

```{r Heat map S103}
# S103

# Create the annotation data frame
annotation_S103 <- metadata %>% 
                    filter (Genotype == "S103") %>% 
                    select(c(Strain,GenPop))

rownames(annotation_S103) <- annotation_S103$Strain
annotation_S103$Strain <- NULL

matrix_S103 <- as.matrix(snps_S103)
nums_S103 <- apply(matrix_S103, c(1, 2), function(x) sprintf("%d", x))

# Plot the heatmap using pheatmap
pheatmap(matrix_S103,
         cluster_rows = TRUE,
         cluster_cols = TRUE,
         display_numbers = nums_S103, 
         color = colorRampPalette(c("yellow", "purple"))(100),
         main = "SNPs heatmap of the strain S103",
         annotation_col = annotation_S103,
         annotation_row = annotation_S103,
         annotation_colors = list_colours)

png(paste0(visuals_path,"/Lac_mutations_heat_S103.png"), width = 4000, height = 3000, res = 300)
pheatmap(matrix_S103,
         cluster_rows = TRUE,
         cluster_cols = TRUE,
         display_numbers = nums_S103, 
         color = colorRampPalette(c("yellow", "purple"))(100),
         main = "SNPs heatmap of the strain S103",
         annotation_col = annotation_S103,
         annotation_row = annotation_S103,
         annotation_colors = list_colours)
dev.off()

```


```{r Genes presence-absence plot S239}

# Load the genes presence-absence matrix and set the gene names as row names
accessory_S239 <- read.table(accessory_path, header = TRUE, check.names = FALSE) %>% 
                                                          column_to_rownames("Gene") %>%
                                                          select(all_of(isolates_S239)) %>% 
                                                          filter(rowSums(. == 1) != ncol(.)) %>% 
                                                          filter(rowSums(. == 0) != ncol(.))
matrix_accessory_S239 <- as.matrix(accessory_S239)

# Plot the heatmap using pheatmap
pheatmap(matrix_accessory_S239, 
         cluster_rows = TRUE, 
         cluster_cols = TRUE, 
         color = c("white", "black"),
         main = "Accessory genes heatmap of the strain S239",
         annotation_col = annotation_S239,
         annotation_colors = list_colours,
         legend = FALSE,
         show_rownames = FALSE)

png(paste0(visuals_path,"/Lac_accessory_heat_S239.png"), width = 4000, height = 3000, res = 300)
pheatmap(matrix_accessory_S239, 
         cluster_rows = TRUE, 
         cluster_cols = TRUE, 
         color = c("white", "black"),
         main = "Accessory genes heatmap of the strain S239",
         annotation_col = annotation_S239,
         annotation_colors = list_colours,
         legend = FALSE,
         show_rownames = FALSE)
dev.off()

# We also calculate the pairwise distance between genomes
pairwise_matrix_accessory_S239 <- as.matrix(dist(t(matrix_accessory_S239), method = "manhattan"))
nums_accessory_S239 <- apply(pairwise_matrix_accessory_S239, c(1, 2), function(x) sprintf("%d", x))

pheatmap(pairwise_matrix_accessory_S239, 
         cluster_rows = TRUE, 
         cluster_cols = TRUE, 
         display_numbers = nums_accessory_S239, 
         color = colorRampPalette(c("yellow", "purple"))(100),
         main = "Accessory genes heatmap of the strain S239",
         annotation_col = annotation_S239,
         annotation_colors = list_colours,
         legend = FALSE,
         show_rownames = FALSE)

```

```{r Genes presence-absence plot S103}

# Load the genes presence-absence matrix and set the gene names as row names
accessory_S103 <- read.table(accessory_path, header = TRUE, check.names = FALSE) %>% 
                                                          column_to_rownames("Gene") %>%
                                                          select(all_of(isolates_S103)) %>% 
                                                          filter(rowSums(. == 1) != ncol(.)) %>% 
                                                          filter(rowSums(. == 0) != ncol(.))
matrix_accessory_S103 <- as.matrix(accessory_S103)

# Plot the heatmap using pheatmap
pheatmap(matrix_accessory_S103, 
         cluster_rows = TRUE, 
         cluster_cols = TRUE, 
         color = c("white", "black"),
         main = "Accessory genes heatmap of the strain S103",
         annotation_col = annotation_S103,
         annotation_colors = list_colours,
         legend = FALSE,
         show_rownames = FALSE)

png(paste0(visuals_path,"/Lac_accessory_heat_S103.png"), width = 4000, height = 3000, res = 300)
pheatmap(matrix_accessory_S103, 
         cluster_rows = TRUE, 
         cluster_cols = TRUE, 
         color = c("white", "black"),
         main = "Accessory genes heatmap of the strain S103",
         annotation_col = annotation_S103,
         annotation_colors = list_colours,
         legend = FALSE,
         show_rownames = FALSE)
dev.off()

# We also calculate the pairwise distance between genomes
pairwise_matrix_accessory_S103 <- as.matrix(dist(t(matrix_accessory_S103), method = "manhattan"))
nums_accessory_S103 <- apply(pairwise_matrix_accessory_S103, c(1, 2), function(x) sprintf("%d", x))

pheatmap(pairwise_matrix_accessory_S103, 
         cluster_rows = TRUE, 
         cluster_cols = TRUE, 
         display_numbers = nums_accessory_S103, 
         color = colorRampPalette(c("yellow", "purple"))(100),
         main = "Accessory genes heatmap of the strain S103",
         annotation_col = annotation_S103,
         annotation_colors = list_colours,
         legend = FALSE,
         show_rownames = FALSE)

```




```{r}


# Load the MSA data
aln <- readDNAStringSet(aln_path, format = "fasta")
aln_S239 <- as.matrix(aln[names(aln) %in% isolates_S239])
aln_S239 <- aln_S239[, !apply(aln_S239, 2, function(column) length(unique(column)) == 1)]
aln_S239 <- as.data.frame(t(aln_S239))
aln_S239$Position <- 1:nrow(aln_S239)

# Melt the data frame for ggplot2
aln_S239_long <- melt(aln_S239, id.vars = "Position", variable.name = "Genome", value.name = "Nucleotide")

# Define a custom color palette for nucleotides
nucleotide_colors <- c(
  "A" = "green",
  "C" = "blue",
  "G" = "yellow",
  "T" = "red",
  "-" = "grey"  # For gaps
)


# Create the plot
ggplot(aln_S239_long, aes(x = Position, y = Genome, fill = Nucleotide)) +
  geom_tile() +
  scale_fill_manual(values = nucleotide_colors) +
  labs(title = "Multiple Sequence Alignment", x = "Position", y = "Genome") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))


```


