---
title: "Lac_strains_pangenome"
author: "Bosco Gracia Alvira"
date: "2025-02-24"
output: html_document
---


```{r Load libraries}

# Standard table management
library(data.table) ; packageVersion("data.table")
library(tidyverse) ; packageVersion("tidyverse")
library(readr) ; packageVersion("readr")
library(dplyr) ; packageVersion("dplyr")

# Phylogenetic tree visualisation
library(treeio) ; packageVersion("treeio")
library(ggtree) ; packageVersion("ggtree")
library(phytools) ; packageVersion("phytools")

# Plot the pangenome graphs
library(pheatmap) ; packageVersion("pheatmap")
library(ggplot2) ; packageVersion("ggplot2")
library(tidygraph) ; packageVersion("tidygraph")

```


```{r Set paths}

metadata_path <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Microbiome_pangenomic_analysis/data/Lactiplantibacillus_plantarum/Pangenomic_strains/metadata.tsv"
tree_path <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Microbiome_pangenomic_analysis/data/Lactiplantibacillus_plantarum/Pangenomic_strains/snp-sites/Lpla.phylip.treefile"
pan_path <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Microbiome_pangenomic_analysis/data/Lactiplantibacillus_plantarum/Pangenomic_strains/roary/gene_presence_absence.Rtab"
ANI_path <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Microbiome_pangenomic_analysis/data/Lactiplantibacillus_plantarum/Pangenomic_strains/pyANI/ANIb_percentage_identity.tab"
visuals_path <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Microbiome_pangenomic_analysis/visuals"

```

```{r Phylogenetic tree plot}
  
# Load the metadata file and set the Strains as key
metadata <- fread(metadata_path)
row.names(metadata) <- metadata$Strain
setkey(metadata, Strain)

# Load the representative genomes tree based on core genome SNPs
tree <- read.tree(tree_path)
tree <- ladderize(tree, right = FALSE)

# Re-root the tree to the midpoint between the outgroup and the rest of the tree
tree <- midpoint.root(tree)

# Build the tree plot with bootstrap values and metadata
tree_plot <- ggtree(tree, size = 0.3) %<+% metadata + 
                  geom_tippoint(aes(color = Temperature), size = 0.5) +
                  geom_tiplab(aes(color = Temperature), size = 4, align = TRUE) +
                  geom_text2(aes(subset = !isTip, label = label), hjust = 1.2, vjust = -0.5, size = 2) +
                  scale_color_manual(values = c("HOT" = "red", "COLD"="blue", "BASE"="grey")) +
                  theme_tree2(plot.margin=unit(c(0,1,0,0), 'cm')) +
                  scale_y_continuous(expand = c(0.05, 0.20))
tree_plot

# Save the order of the genomes in the tree
genome_order <- tree_plot$data %>%
                    filter(isTip) %>%
                    arrange(y) %>%
                    pull(label)

```

```{r Genes presence-absence plot}

# Load the genes presence-absence matrix and set the gene names as row names
pan <- read.table(pan_path, header = TRUE, check.names = FALSE) 
pan_matrix <- as.matrix(pan[,-1])
rownames(pan_matrix) <- pan$Gene

# Perform hierarchical clustering on the genes
gene_dist <- dist(pan_matrix)
gene_clust <- hclust(gene_dist)

# Reorder the genes based on the clustering results and the rows based on phylogeny
gene_order <- gene_clust$order
matrix_ordered <- pan_matrix[gene_order,]

pan_long <- as.data.frame(matrix_ordered) %>% 
           rownames_to_column(var = "Gene") %>%
           pivot_longer(cols = -Gene, names_to = "Genome", values_to = "Presence") %>%
           mutate(Presence = as.factor(Presence))

pan_plot <- ggplot(pan_long, aes(x = Genome, y = Gene, fill = Presence)) +
                  geom_tile() +
                  scale_fill_manual(values = c("0" = "white", "1" = "blue")) +
                  theme_minimal() +
                  theme(axis.text.x = element_blank(),
                      axis.ticks.x = element_blank(),
                      axis.text.y = element_text(angle = 0, hjust = 1)) +
                  labs(title = "Gene Presence/Absence Matrix", x = "Genome", y = "Gene") +
                  coord_flip()

pan_plot

```


```{r ANI plot}

# Read the ANI data
ANI <- read.table(ANI_path, header = TRUE, check.names = FALSE) 

# Transform the data into a matrix
ANI_matrix <- as.matrix(ANI[,-1])

# Plot the heatmap using pheatmap
pheatmap(ANI_matrix, 
         cluster_rows = TRUE, 
         cluster_cols = TRUE, 
         display_numbers = FALSE, 
         color = colorRampPalette(c("white", "blue"))(100),
         main = "Average Nucleotide Identity (ANI) Heatmap")

```


```{r Combined plot}

# We have plotted the clustering of the genomes by three independent analyses.
# Now we want to make a combined plot by parsing the presence-absence and ANI info to the phylogenetic tree
gene_list <- data.frame(Gene = row.names(matrix_ordered), Position = seq_along(row.names(matrix_ordered)))

combined_pan <- pan_long %>%
                    left_join(gene_list, by = "Gene") %>% 
                    filter(Presence == 1) %>% 
                    select(-Presence,-Gene)

combined_tree_pan_plot <- tree_plot + 
                              geom_facet(panel = "Gene presence-absence", 
                                         data = combined_pan, 
                                         geom = geom_point, 
                                         mapping = aes(x = Position), 
                                         shape = '|', size=4) +
                              theme(axis.title.x = element_blank(),  # Remove x-axis title
                                    axis.text.x = element_blank(),   # Remove x-axis labels
                                    axis.ticks.x = element_blank(),  # Remove x-axis ticks
                                    panel.grid.major.x = element_blank(),  # Remove major grid lines
                                    panel.grid.minor.x = element_blank(),
                                    plot.margin=unit(c(0,1,0,0), 'cm')) +
                              theme_void()

ANI_list <- data.frame(Genome2 = genome_order, Position = seq_along(genome_order))

combined_ANI <- as.data.frame(ANI) %>% 
                       rownames_to_column(var = "Genome") %>%
                       pivot_longer(cols = -Genome, names_to = "Genome2", values_to = "ANI") %>%
                       mutate(ANI = as.numeric(ANI)*100) %>% 
                       left_join(ANI_list, by = "Genome2") %>% 
                       select(-Genome2)

combined_tree_pan_ANI_plot <- combined_tree_pan_plot +
                                  geom_facet(panel = "ANI", 
                                             data = combined_ANI, 
                                             geom = geom_tile, 
                                             mapping = aes(x = Position, fill = ANI)) +
                                  theme(aspect.ratio = 1) + 
                                             scale_fill_gradient(low = "purple", high = "yellow") +
                                  theme(axis.title.x = element_blank(),
                                        axis.text.x = element_blank(),
                                        axis.ticks.x = element_blank(),
                                        panel.grid.major.x = element_blank(),
                                        panel.grid.minor.x = element_blank(),
                                        plot.margin=unit(c(0,0,0,0), 'cm'))

combined_tree_pan_ANI_plot

ggsave(filename = "Lac_strains_combined.png", plot = combined_tree_pan_ANI_plot, path = visuals_path, width = 25, height = 10, dpi = 300)

```

```{r Fisher test}

# We want to test for association between genotype and source of isolation of the isolates. In particular, we want to test if there is a significant association between the temperature regime of isolation and the genotype of the strain.

# Load the metadata file
metadata_fisher <- metadata %>% 
                          mutate(Temperature_random = sample(Temperature),
                                 Population_random = sample(Population),
                                 Generation_random = sample(Generation))

# List of parameters to test
parameters <- c("Temperature", "Population", "Generation")

# Initialize a list to store the results
results <- list()

# Perform Fisher's exact tests for each parameter
for (param in parameters) {
  contingency_table <- table(metadata$Genotype, metadata[[param]])
  fisher_test <- fisher.test(contingency_table)
  results[[param]] <- fisher_test$p.value
}

# Convert the results to a data frame
results_df <- data.frame(Parameter = names(results), P_Value = unlist(results))

# Apply Bonferroni correction
results_df$Adjusted_P_Value <- p.adjust(results_df$P_Value, method = "bonferroni")

# Print the results
print(results_df)

```